Added new metric type Event.

GitAtSmsPortal opened the pull request:
Added cleaner to clean out events after a period of time after they have been reported.
Updated the MetricTags GetHashCode function to be public static so that we can get the hash for a strin[] of tags in multilpe places.
Updated the reportes to report the event metrics.

Commit:
Added new metric type Event.
Added cleaner to clean out events after a period of time after they have been reported.
Updated the MetricTags GetHashCode function to be public static so that we can get the hash for a strin[] of tags in multilpe places.
Updated the reportes to report the event metrics.

Commit:
Replaced tabs with spaces.

Commit:
Fixed deregister bug to take into account UseMetricTypeIdentifiers.
Removed silly random characters in class.

Commit:
Added helper class to manages the generation of the unique identifier of a metric.

Commit:
Added code to remove the event from the cleaners internal collection when deregister is called.

Commit:
no message

Commit:
no message

Commit:
Added unit tests for event metric and cleaner.

Commit:
Added code to clear the event details when no reports are register and when all reports are filtering out the events.

Francois-du-Plessis commented on Src/Metrics.Tests/Reporting/EventMetricsCleanerTest.cs:
```diff
@@ -0,0 +1,212 @@
+        [Fact]
+        public void StartsWithZeroReports()
+        {
+            EventMetricsCleaner.Clear();
```
Clear should not be called here.

GitAtSmsPortal replied:
Right you are. Have made all other tests clear after they have run.

Francois-du-Plessis commented on Src/Metrics.Tests/Reporting/EventMetricsCleanerTest.cs:
```diff
@@ -0,0 +1,212 @@
+        }
+
+        [Fact]
+        public void RegisterReport_WithGreaterTimespan_UpdatesIntervalToBeGreaterThanReportInterval()
```
Could test WithSmallerTimespam_DoesNotUpdateInterval

GitAtSmsPortal replied:
Have added a unit test to test the correct logic.

Commit:
Moved clear method to run after test has run.

Francois-du-Plessis commented on Src/Metrics.Tests/Reporting/EventMetricsCleanerTest.cs:
```diff
@@ -0,0 +1,212 @@
+
+            timer.OnTimerCallback();
+
+            EventMetricsCleaner.GetEventDetailCount("test").Should().Be(1);
```
Should this not rather look at the event list in the DefaultMetricsRegistry to see if it actually clears the event

Commit:
Added a reset interval function to the cleaner.
Updated unit tests to reset the intervals after each test runs.
Added unit test to check largest report interval is used when registering a report.

Commit:
no message

Commit:
no message
